vrclient_x64/unix_vrcompositor_manual.cpp

```diff
@@ -27,31 +27,45 @@ static u_VRVulkanTextureArrayData_t *unwrap_texture_vkdata_array( const w_VRVulk
 }
 
 static u_VRTextureWithPoseAndDepth_t unwrap_texture_with_pose_and_depth( const w_VRTextureWithPoseAndDepth_t *w_texture, uint32_t flags, u_VRVulkanTextureArrayData_t *u_vkdata,
-                                                                         u_VRVulkanTextureData_t *u_depth_vkdata )
+                                                                         u_VRVulkanTextureArrayData_t *u_depth_vkdata )
 {
     u_VRTextureWithPoseAndDepth_t u_texture = *w_texture;
 
     if (w_texture->eType == TextureType_Vulkan)
     {
-        if (flags & Submit_VulkanTextureWithArrayData) u_texture.handle = unwrap_texture_vkdata_array( (w_VRVulkanTextureArrayData_t *)w_texture->handle, u_vkdata );
-        else u_texture.handle = unwrap_texture_vkdata( (w_VRVulkanTextureData_t *)w_texture->handle, (u_VRVulkanTextureData_t *)u_vkdata );
-        u_texture.depth.handle = unwrap_texture_vkdata( (w_VRVulkanTextureData_t *)w_texture->depth.handle, u_depth_vkdata );
+        if (flags & Submit_VulkanTextureWithArrayData)
+        {
+            u_texture.handle = unwrap_texture_vkdata_array( (w_VRVulkanTextureArrayData_t *)w_texture->handle, u_vkdata );
+            u_texture.depth.handle = unwrap_texture_vkdata_array( (w_VRVulkanTextureArrayData_t *)w_texture->depth.handle, u_depth_vkdata );
+        }
+        else
+        {
+            u_texture.handle = unwrap_texture_vkdata( (w_VRVulkanTextureData_t *)w_texture->handle, (u_VRVulkanTextureData_t *)u_vkdata );
+            u_texture.depth.handle = unwrap_texture_vkdata( (w_VRVulkanTextureData_t *)w_texture->depth.handle, (u_VRVulkanTextureData_t *)u_depth_vkdata );
+        }
     }
 
     return u_texture;
 }
 
 static u_VRTextureWithDepth_t unwrap_texture_with_depth( const w_VRTextureWithDepth_t *w_texture, uint32_t flags, u_VRVulkanTextureArrayData_t *u_vkdata,
-                                                         u_VRVulkanTextureData_t *u_depth_vkdata )
+                                                         u_VRVulkanTextureArrayData_t *u_depth_vkdata )
 {
     u_VRTextureWithDepth_t u_texture = *w_texture;
 
     if (w_texture->eType == TextureType_Vulkan)
     {
-        if (flags & Submit_VulkanTextureWithArrayData) u_texture.handle = unwrap_texture_vkdata_array( (w_VRVulkanTextureArrayData_t *)w_texture->handle, u_vkdata );
-        else u_texture.handle = unwrap_texture_vkdata( (w_VRVulkanTextureData_t *)w_texture->handle, (u_VRVulkanTextureData_t *)u_vkdata );
-        /* We should maybe unwrap the vkdata but No Man Sky uses a garbage handle in its w_VRTextureDepthInfo_t, is this really used? */
-        u_texture.depth.handle = w_texture->depth.handle;
+        /* No Man Sky uses a garbage handle in its w_VRTextureDepthInfo_t, is depth without pose really used? */
```

Comment:
What about that comment now? Is this something that was missing from the DXVK unwrap, or is it still garbage? If this is still garbage we can't safely access it.

Reply:
That is a good question. I don't have _No Man Sky_ (_No Man's Sky_?) to test if it works or not. I did build build a [`redist` proton with these commits](https://staff.sharcnet.ca/tyson/experimental-bleeding-edge-8.0-65469-20231124-pb6bf0c-w34e4e3-d91f7f4-veb4b41-2-g73e569ac.tar.gz) though that someone could test.

I can see a couple of options
1. _No Man Sky_ works, so we are good as is.
2. Let the developers of _No Man Sky_ know they are incorrectly using `Submit_TextureWithDepth` as they are not submitting with a depth texture so they fix it.
3. If it is possible to determine if the application is _No Man Sky_ (an if on the SteamAppId or something?), make the unpack conditional on that.
4. Don't unpack the just depth case.

This last one would still unpack Just pose or pose and depth (the two cases I am actually interested in), so, from my perspective, it would be okay. Although it does seem kind of wrong as it would then be broken for any other application that actually does use `Submit_TextureWithDepth` properly.

Reply:
No problem. Thanks very much for looking at it. :+1:

Reply:
> What about that comment now? Is this something that was missing from the DXVK unwrap, or is it still garbage? If this is still garbage we can't safely access it.

No Man's Sky is a Vulkan game, no dxvk unwrap is used. I think it still passes broken data here (at least we had seen that a few weeks ago I guess), and it is still not used by the compositor. I suppose only the first patch is needed for the games to work? I think we can leave that without unwrapping (with some warn maybe), but we can't unconditionally unwrap.

Reply:
My reading of the code was that the it all goes through the same vulkan code path. There is just an if statement for `TextureType_DirectX` that adds adds a [pre-vulkan-texture-extraction-and-optimal-translation call before](https://github.com/ValveSoftware/Proton/blob/d6b3aa9/vrclient_x64/vrcompositor_manual.c#L1298) and a reverse-translation on the way out.

I'll do up a further commit for the depth only case. Maybe a static local variable and a warning so people can report if they run into it in any other programs.

Reply:
What I am trying to say that you probably don't really need to path depth texture anyhow. What is actually fixing the game is passing pose and handling the flag combination. Can you please check if the game works if instead of unwrapping depth texture you just pass the null handle it still works? I'd expect it so, and then we could get rid of most of the code in those patches, also avoiding breaking NMS on the way.

Reply:
On a separate note, are you checking this with Steam VR or some custom compositor?

Reply:
The particular application I was looking at was [Revive](https://github.com/LibreVR/Revive). It is a dll that you inject into an Oculus application. It hooks the Oculus API calls and translates them to OpenVR calls instead. There is a [fairly extensive list](https://github.com/LibreVR/Revive/wiki/Compatibility-list) of titles that it works for.

Right now it just does `Submit_TextureWithPose`. The author (@CrossVR) [said](https://github.com/LibreVR/Revive/issues/1571#issuecomment-1819224854)
> The flag is necessary to ensure stable reprojection. To closely align with the Oculus API Revive does its own pose prediction for each frame. The pose that was used to render the frame is then sent to the compositor to use for reprojection.

There is also a [`ENABLE_DEPTH_SUBMIT` define](https://github.com/LibreVR/Revive/blob/ab73167/Revive/CompositorBase.cpp#L423) in the code that will cause it to do `Submit_TextureWithDepth` too . Currently this isn't enabled by default. I don't know if it will be at some point or not. I figured I might as well do them all while I was at it though as the hard part was getting up to speed on the code, getting it accepted, etc., not doing the actual patch.

With regard to other possible titles, googling for _fixme:vrclient:ivrcompositor_submit_dxvk Unhandled flags 0x8_ (`Submit_TextureWithPose`) gives reports that mention _L.A. Noire: The VR Case Files_ and various VR mods (_GTA5 VR_, _Red Dead Redemption 2 VR_, _CyberPunk 2077_). Doing the same for flags _0x10_ (`Submit_TextureWithDepth`) and  _0x18_ (`Submit_TextureWithPose | Submit_TextureWithDepth`) doesn't return any hits.

I am just using the Steam VR compositor. From what you are saying, it even if `Revive` did do `Submit_TextureWithDepth` at some point, it wouldn't matter as that isn't used.

Reply:
Thanks!

So if I understand correctly, from what you are saying nothing really depends on actually unwrapping the textures. Also, as I noted, SteamVR seems to ignore depth texture anyway, doesn't matter whether the application passes that or not.

So could you please remove all the code related to actual texture unwrapping from PR and just handle flags and their combination, along with correctly passing the pose? Submit_TextureWithDepth and both pose in depth can be handled, you just don't need to actually mind the texture. That will remove the majority of most complicated and regression prone code from the MR. Then I'd look at the final version in more details for nits and we could get that in.

If SteamVR gets actual support for depth texture we might get back to correctly delivering the data, before that, even regardless of NMS, it doesn't make much sense to add effectively dead code which we can't even test.